pkg/kv/kvserver/loqrecovery/server.go

```diff
@@ -0,0 +1,216 @@
+}
+
+func wrapRetryable(err error) error {
+	return &retryVisit{cause: err}
+}
+
+func IsRetryError(err error) (bool, error) {
+	if r := (*retryVisit)(nil); errors.As(err, &r) {
+		return true, r.cause
+	}
+	return false, err
+}
+
+// ClusterAdminClient allows Server to talk to all remote nodes without having
+// a dependence on Admin Server.
+type ClusterAdminClient interface {
+	// VisitAvailableNodesWithRetry calls visitor with an admin client for every
+	// known cluster node.
+	// If visitor returns retryVisit error then node would be dialed again and
+	// revisited until visitor succeeds or retry attempts are exhausted. In latter
+	// case cause of retryVisit is returned.
+	// If visitor returns StopIteration() error, rest of the nodes would be
+	// skipped.
+	// Returns error if visitor failed or if error happened when processing node
+	// data.
+	VisitAvailableNodesWithRetry(ctx context.Context, retryOpts retry.Options,
+		visitor func(nodeID roachpb.NodeID, client serverpb.AdminClient) error,
+	) error
+}
+
+type Server struct {
+	stores               *kvserver.Stores
+	metadataQueryTimeout time.Duration
+	forwardReplicaFilter func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
+}
+
+func NewRecoveryServer(stores *kvserver.Stores, knobs base.ModuleTestingKnobs) *Server {
+	metadataQueryTimeout := 1 * time.Minute
+	var forwardReplicaFilter func(*serverpb.RecoveryCollectLocalReplicaInfoResponse) error
+	if rk, ok := knobs.(*TestingKnobs); ok && rk.MetadataScanTimeout > 0 {
+		metadataQueryTimeout = rk.MetadataScanTimeout
+		forwardReplicaFilter = rk.ForwardReplicaFilter
+	}
+	return &Server{
+		stores:               stores,
+		metadataQueryTimeout: metadataQueryTimeout,
+		forwardReplicaFilter: forwardReplicaFilter,
+	}
+}
+
+func (s Server) ServeLocalReplicas(
+	ctx context.Context,
+	_ *serverpb.RecoveryCollectLocalReplicaInfoRequest,
+	stream serverpb.Admin_RecoveryCollectLocalReplicaInfoServer,
+) error {
+	return s.stores.VisitStores(func(s *kvserver.Store) error {
+		reader := s.Engine().NewSnapshot()
+		defer reader.Close()
+		return collectReplicasFromStore(ctx, reader, s.StoreID(), s.NodeID(),
+			func(info loqrecoverypb.ReplicaInfo) error {
+				return stream.Send(&serverpb.RecoveryCollectLocalReplicaInfoResponse{ReplicaInfo: &info})
+			})
+	})
+}
+
+func (s Server) ServeClusterReplicas(
+	ctx context.Context,
+	_ *serverpb.RecoveryCollectReplicaInfoRequest,
+	outStream serverpb.Admin_RecoveryCollectReplicaInfoServer,
+	kvDB *kv.DB,
+	clusterAdminClient ClusterAdminClient,
+) (err error) {
+	var (
+		descriptors, nodes, replicas int
+	)
+	defer func() {
+		if err == nil {
+			log.Infof(ctx, "streamed info: range descriptors %d, nodes %d, replica infos %d", descriptors,
+				nodes, replicas)
+		}
+	}()
+
+	err = contextutil.RunWithTimeout(ctx, "scan-range-descriptors", s.metadataQueryTimeout,
+		func(txnCtx context.Context) error {
+			txn := kvDB.NewTxn(txnCtx, "scan-range-descriptors")
```

Comment:
We should look closer at the implications of this transaction, to ensure it doesn't cause problems with concurrent writes. We've seen indications elsewhere (e.g. below-Raft migrations) that long-running meta scans can be problematic.

Reply:
If we're certain that this doesn't content with concurrent writers then it should be fine.